external-crates/move/crates/move-compiler/src/linters/redundant_boolean_expressions.rs

```diff
@@ -0,0 +1,91 @@
+//! This rules detects and warns about unnecessary boolean operation such as x || true.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    expansion::ast::Value_,
+    naming::ast::Var_,
+    parser::ast::BinOp_,
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::{
+        ast::{self as T, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+use super::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+const REDUNDANT_BOOLEAN_EXP_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::RedundantBooleanExp as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "",
+);
+
+pub struct RedundantBooleanExp;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for RedundantBooleanExp {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::BinopExp(e1, op, _, e2) = &exp.exp.value {
+            // Check if the operation is a logical OR
+            if let BinOp_::Or = &op.value {
+                match (&e1.exp.value, &e2.exp.value) {
```

Comment:
We should really handle any and all redundant boolean expressions, and merge this with 
#16488